### ISSUE TITLE
Maintenance of this crate

### ISSUE THREAD

**Issue by adeschamps** [adeschamps/lcm#24]
I wouldn't say it's the case that I don't have time to maintain this crate. However, I'm not working on any projects that make use of it, and therefore I don't really have a sense of how well it's working.

@neachdainn Basically what I'm saying is: I think you're the only person currently using this, and so if you want to make breaking changes, go for it!

**Comment by neachdainn** [comment]
I believe you'll have to still push the updates to Crates.io but otherwise sounds good.

**Comment by neachdainn** [comment]
For the record, I am still planning things for this crate. I'm trying to get my head wrapped around Tokio to create a pure Rust implementation of LCM.

**Comment by adeschamps** [comment]
What's your motivation for wanting a pure Rust implementation?

I've been thinking about rewriting the codegen in Rust so that this can be used without having to build lcm-gen from the fork.

**Comment by neachdainn** [comment]
There isn't one big reason, just several small ones. E.g., handling panics across FFI boundaries. The latest published version of the crate does nothing (which leads to undefined behavior) and the current version in the repository just discards the panic. Fixing that would require building some kind of mechanism to resume the panic once execution is back in Rust code. 

Another example would be paying the cost of being thread-safe even in a single threaded environment. If this was a pure Rust implementation, it would be possible to avoid locking/unlocking a mutex on every message, which is what I believe the underlying C code does. **EDIT:** We would be able to avoid the mutex in the single-threaded case while still having options for the multi-threaded.

I also really just want to figure out Tokio.

**Comment by neachdainn** [comment]
I've also been thinking about rewriting the codegen in Rust for two major reasons:

1. Trying to make all the changes that I would like to see will be annoying to do in C.
2. Having a pure-Rust implementation of the codegen is the first step towards having a custom derive.

**Comment by adeschamps** [comment]
I think that reimplementing LCM itself and reimplementing the codegen can probably be done more or less independently.

I originally wrote this crate as a means of learning Rust. Panics are definitely not handled correctly, and there are probably other significant issues as well.

Do you have a use case for multithreaded LCM? I my own projects, I might have multiple instances of LCM running in different threads, but I've never had a case where I needed to share a single LCM instance across multiple threads.

I've made several attempts at rewriting the codegen. I wrote a nom parser at one point, but it didn't handle comments. More recently I tried using [pest], which was really easy to get started, although I've run into troubles by writing a grammar that resulted in an infinite loop. **[edit: I jus revisited it, and I figured it out.]**

I'll create some new issues, since this conversation seems to be diverging into several directions.

[pest]: https://docs.rs/pest/1.0.3/pest/

**Comment by adeschamps** [comment]
Oh, and if we decide to do both of those things, then I don't think it makes sense for this to be a fork of the official repo anymore. In that case, I propose we create a new repo called `lcm-rust` or something similar.

**Comment by neachdainn** [comment]
I agree with creating the new repo. I'll comment more on the other things in their respective issues.
